### PR TITLE
2023-09-10 fizzbuzz attempt 1

### PR DESCRIPTION
**Pull request by daneel-deimos**
# Grading Checklist
- [x] I have committed on every single transition from red to green to refactor
- [x] I have tests that validate the following statements
1. each return value is a string 
2. 3 returns "Fizz"
3. 5 returns "Buzz
4. 15 returns "FizzBuzz"
5. 9 returns "Fizz"
6. 43 returns "43"
7. 42 returns "Fizz"
8. 45 returns "FizzBuzz"
9. 102 (you decide, throw an Error or handle some other way)
10. -12 (you decide, throw an Error or handle some other way)
11. any non-number (you decide, throw an Error or handle some other way)
- [x] Once I have made the aforementioned tests pass, I have refactored my test specifications to use it.each() to perform parameterization
- [x] There is no duplication in my test code or my production code

### CODE REVIEW

File: 1_The_Physical/Getting_Started_Emergent_Design/1_Red_Green_Refactor/1_1_Fizzbuzz/src/index.spec.ts [hunk 1]
```diff
@@ -16,12 +16,10 @@ describe("fizzbuzz", () => {
...
-    it('should return "FizzBuzz" when provided with the value 15', () => {
-        expect(fizzBuzz(15)).toBe('FizzBuzz');
-    })
-    
-    it('should return "FizzBuzz" when provided with the value 45', () => {
-        expect(fizzBuzz(45)).toBe('FizzBuzz')
+    it('should "FizzBuzz" when provided with the values 15 and 45', () => {
+        [15, 45]
+            .map((input:number) => fizzBuzz(input))
+            .forEach((output) => expect(output).toBe('FizzBuzz'));
```
Comment (taha-crafter) [inline 1_The_Physical/Getting_Started_Emergent_Design/1_Red_Green_Refactor/1_1_Fizzbuzz/src/index.spec.ts]: you could get rid of the `map `loop, it's an unnecessary loop here, you can just use `forEach`, like this:
```ts
it('should return "FizzBuzz" when provided with the values 15 and 45', () => {
  [15, 45].forEach((input) => {
    expect(fizzBuzz(input)).toBe('FizzBuzz');
  });
});
```

**Review by taha-crafter — COMMENTED** [review]
Hi @daneel-deimos,

Great job on your implementation! 👍🏼 

I liked how you stuck to the Fake it strategy :) you stuck to it to the point where your end result code only works for the examples in the grading checklist :) 
```ts
 if (input === 15 || input === 45) {
        return 'FizzBuzz';
    }
```
You didn't refactor to an Obvious Implementation because in the grading checklist examples, there are only 2 examples for cases of FizzBuzz and we're asked to only refactor to Obvious Implementation after seeing 3 repetitions in the code.

While sticking to the Fake it strategy is a good approach, it's important to ensure that your solution works for all cases described in the problem statement, not just the examples provided in the grading checklist.

By adding additional examples, you can verify that your code handles various scenarios correctly, such as returning "FizzBuzz" for any multiples of both 3 and 5. This will give you more confidence in the accuracy and robustness of your solution.

Although the grading checklist may not explicitly mention adding extra examples, it's generally a good practice to test your code with different inputs to cover a wide range of scenarios. This helps in catching any potential edge cases or unexpected behavior.

Consider incorporating additional test cases that cover different combinations of multiples of 3 and 5. This will ensure that your solution is comprehensive and handles all possible inputs as per the problem statement.

Keep up the good work and continue refining your implementation to make it more versatile and reliable!

File: 1_The_Physical/Getting_Started_Emergent_Design/1_Red_Green_Refactor/1_1_Fizzbuzz/src/index.spec.ts (hunk 1 above)
Comment (daneel-deimos) [reply 1_The_Physical/Getting_Started_Emergent_Design/1_Red_Green_Refactor/1_1_Fizzbuzz/src/index.spec.ts]: Resolved. Thank you for the feedback 🙏🏽

**Comment by daneel-deimos** [comment]
Thank you so much taking the time to review my work [taha-crafter](https://github.com/taha-crafter).

I've reviewed your feedback and incorporated it in two follow up commits. I've added additional test cases and refactored the tests themselves.

Much appreciated.
